triedb/pathdb/reader.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>
+
+package pathdb
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/triedb/database"
+)
+
+// The types of locations where the node is found.
+const (
+	locDirtyCache = "dirty" // dirty cache
+	locCleanCache = "clean" // clean cache
+	locDiskLayer  = "disk"  // persistent state
+	locDiffLayer  = "diff"  // diff layers
+)
+
+// nodeLoc is a helpful structure that contains the location where the node
+// is found, as it's useful for debugging purposes.
+type nodeLoc struct {
+	loc   string
+	depth int
+}
+
+// string returns the string representation of node location.
+func (loc *nodeLoc) string() string {
+	return fmt.Sprintf("loc: %s, depth: %d", loc.loc, loc.depth)
+}
+
+// reader implements the Reader interface, providing the functionalities to
+// retrieve trie nodes by wrapping the internal state layer.
+type reader struct {
+	layer layer
+	state crypto.KeccakState
+}
+
+// Node implements trie.Reader interface, retrieving the node with specified
+// node info. Don't modify the returned byte slice since it's not deep-copied
+// and still be referenced by database.
+func (r *reader) Node(owner common.Hash, path []byte, hash common.Hash) ([]byte, error) {
+	blob, loc, err := r.layer.node(owner, path, 0)
+	if err != nil {
+		return nil, err
+	}
+	if got := crypto.HashData(r.state, blob); got != hash {
```

Comment:
This is dangerous. You are using `r.state` without mutex protection. I got a few different errors
```
[user@work go-ethereum]$ go test ./eth/protocols/snap -run TestSyncNoStorageAndOneCappedPeer
panic: runtime error: index out of range [100] with length 0

goroutine 510 [running]:
golang.org/x/crypto/sha3.(*state).padAndPermute(0xc00047f340, 0x5c?)
        /home/user/go/pkg/mod/golang.org/x/crypto@v0.21.0/sha3/sha3.go:111 +0x237
golang.org/x/crypto/sha3.(*state).Read(0xc00047f340, {0xc0004e0a00, 0x20, 0x20})
        /home/user/go/pkg/mod/golang.org/x/crypto@v0.21.0/sha3/sha3.go:164 +0x45
github.com/ethereum/go-ethereum/crypto.HashData({0x7f25cc4c9218, 0xc00047f340}, {0xc0003e6d80, 0x214, 0x214})
        /home/user/go/src/github.com/ethereum/go-ethereum/crypto/crypto.go:80 +0x8b
github.com/ethereum/go-ethereum/triedb/pathdb.(*reader).Node(0xc00075c000, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
        /home/user/go/src/github.com/ethereum/go-ethereum/triedb/pathdb/reader.go:63 +0xaf
```
Second time: 
```
$ go test ./eth/protocols/snap -run TestSyncNoStorageAndOneCappedPeer
--- FAIL: TestSyncNoStorageAndOneCappedPeer (0.12s)
    sync_test.go:1056: accounts: 3000, slots: 0
    sync_test.go:250: Remote side rejected our delivery: proof node (hash 855ada8d6587d7375a813dfc5bda76300d226f0263eb9435c44a5ca5e1d3bda0) missing
    sync_test.go:250: Remote side rejected our delivery: proof node (hash f3d5fddd9b908502006b4ea926d4b1146e6034e7e531abc91eccb24c20e5d32e) missing
    sync_test.go:1053: sync failed: sync cancelled
panic: Fail in goroutine after TestSyncNoStorageAndOneCappedPeer has completed

goroutine 438 [running]:
testing.(*common).Fail(0xc000384820)
        /usr/local/go/src/testing/testing.go:952 +0xd4
testing.(*common).Errorf(0xc000384820, {0xf3d84b?, 0x155?}, {0xc0007bdf38?, 0x0?, 0x0?})
        /usr/local/go/src/testing/testing.go:1075 +0x5e
github.com/ethereum/go-ethereum/eth/protocols/snap.defaultAccountRequestHandler(0xc00021c0b0, 0x146ae7a6cdf4e2aa, {0x75, 0x65, 0x93, 0x76, 0x76, 0x40, 0x2, 0xcc, ...}, ...)
        /home/user/go/src/github.com/ethereum/go-ethereum/eth/protocols/snap/sync_test.go:250 +0x131
created by github.com/ethereum/go-ethereum/eth/protocols/snap.(*testPeer).RequestAccountRange in goroutine 455
        /home/user/go/src/github.com/ethereum/go-ethereum/eth/protocols/snap/sync_test.go:188 +0x394
FAIL    github.com/ethereum/go-ethereum/eth/protocols/snap      0.203s
```
Such access could happen e.g. thorugh `*trie.Trie.Prove`



Running it with `--race` yields: 
```
==================
WARNING: DATA RACE
Write at 0x00c0005c8540 by goroutine 400:
  golang.org/x/crypto/sha3.(*state).Reset()
      /home/user/go/pkg/mod/golang.org/x/crypto@v0.21.0/sha3/sha3.go:61 +0x44
  github.com/ethereum/go-ethereum/crypto.HashData()
      /home/user/go/src/github.com/ethereum/go-ethereum/crypto/crypto.go:78 +0x59
  github.com/ethereum/go-ethereum/triedb/pathdb.(*reader).Node()
      /home/user/go/src/github.com/ethereum/go-ethereum/triedb/pathdb/reader.go:63 +0x11e
  github.com/ethereum/go-ethereum/trie.(*trieReader).node()
      /home/user/go/src/github.com/ethereum/go-ethereum/trie/trie_reader.go:69 +0x1a9
  github.com/ethereum/go-ethereum/trie.(*Trie).Prove()
      /home/user/go/src/github.com/ethereum/go-ethereum/trie/proof.go:71 +0xb74
  github.com/ethereum/go-ethereum/eth/protocols/snap.createAccountRequestResponse()
...
Previous read at 0x00c0005c8540 by goroutine 398:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:325 +0x0
  golang.org/x/crypto/sha3.copyOutUnaligned()
      /home/user/go/pkg/mod/golang.org/x/crypto@v0.21.0/sha3/xor_unaligned.go:58 +0x46
  golang.org/x/crypto/sha3.(*state).padAndPermute()
      /home/user/go/pkg/mod/golang.org/x/crypto@v0.21.0/sha3/sha3.go:121 +0x492
  golang.org/x/crypto/sha3.(*state).Read()
      /home/user/go/pkg/mod/golang.org/x/crypto@v0.21.0/sha3/sha3.go:164 +0x8c
  github.com/ethereum/go-ethereum/crypto.HashData()
      /home/user/go/src/github.com/ethereum/go-ethereum/crypto/crypto.go:80 +0x98
  github.com/ethereum/go-ethereum/triedb/pathdb.(*reader).Node()
      /home/user/go/src/github.com/ethereum/go-ethereum/triedb/pathdb/reader.go:63 +0x11e
  github.com/ethereum/go-ethereum/trie.(*trieReader).node()
      /home/user/go/src/github.com/ethereum/go-ethereum/trie/trie_reader.go:69 +0x1a9
  github.com/ethereum/go-ethereum/trie.(*Trie).Prove()
      /home/user/go/src/github.com/ethereum/go-ethereum/trie/proof.go:71 +0xb74
  github.com/ethereum/go-ethereum/eth/protocols/snap.createAccountRequestResponse()


also 

Previous write at 0x00c0005c85c8 by goroutine 398:
  golang.org/x/crypto/sha3.(*state).Reset()
      /home/user/go/pkg/mod/golang.org/x/crypto@v0.21.0/sha3/sha3.go:61 +0x44
  github.com/ethereum/go-ethereum/crypto.HashData()
      /home/user/go/src/github.com/ethereum/go-ethereum/crypto/crypto.go:78 +0x59
  github.com/ethereum/go-ethereum/triedb/pathdb.(*reader).Node()
      /home/user/go/src/github.com/ethereum/go-ethereum/triedb/pathdb/reader.go:63 +0x11e
  github.com/ethereum/go-ethereum/trie.(*trieReader).node()
      /home/user/go/src/github.com/ethereum/go-ethereum/trie/trie_reader.go:69 +0x1a9
  github.com/ethereum/go-ethereum/trie.(*Trie).Prove()
      /home/user/go/src/github.com/ethereum/go-ethereum/trie/proof.go:71 +0xb74
  github.com/ethereum/go-ethereum/eth/protocols/snap.createAccountRequestResponse()
      /home/user/go/src/github.com/ethereum/go-ethereum/eth/protocols/snap/sync_test.go:285 +0x96a
  github.com/ethereum/go-ethereum/eth/protocols/snap.defaultAccountRequestHandler()
      /home/user/go/src/github.com/ethereum/go-ethereum/eth/protocols/snap/sync_test.go:248 +0x104
```
and
```
==================
WARNING: DATA RACE
Write at 0x00c0005c8630 by goroutine 400:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:325 +0x0
  golang.org/x/crypto/sha3.(*state).Write()
      /home/user/go/pkg/mod/golang.org/x/crypto@v0.21.0/sha3/sha3.go:147 +0x36b
  github.com/ethereum/go-ethereum/crypto.HashData()
      /home/user/go/src/github.com/ethereum/go-ethereum/crypto/crypto.go:79 +0x78
  github.com/ethereum/go-ethereum/triedb/pathdb.(*reader).Node()
      /home/user/go/src/github.com/ethereum/go-ethereum/triedb/pathdb/reader.go:63 +0x11e
  github.com/ethereum/go-ethereum/trie.(*trieReader).node()
      /home/user/go/src/github.com/ethereum/go-ethereum/trie/trie_reader.go:69 +0x1a9
  github.com/ethereum/go-ethereum/trie.(*Trie).Prove()
      /home/user/go/src/github.com/ethereum/go-ethereum/trie/proof.go:71 +0xb74
  github.com/ethereum/go-ethereum/eth/protocols/snap.createAccountRequestResponse()
...
Previous write at 0x00c0005c8630 by goroutine 398:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:325 +0x0
  golang.org/x/crypto/sha3.copyOutUnaligned()
      /home/user/go/pkg/mod/golang.org/x/crypto@v0.21.0/sha3/xor_unaligned.go:58 +0x46
  golang.org/x/crypto/sha3.(*state).padAndPermute()
      /home/user/go/pkg/mod/golang.org/x/crypto@v0.21.0/sha3/sha3.go:121 +0x492
  golang.org/x/crypto/sha3.(*state).Read()
      /home/user/go/pkg/mod/golang.org/x/crypto@v0.21.0/sha3/sha3.go:164 +0x8c
  github.com/ethereum/go-ethereum/crypto.HashData()
      /home/user/go/src/github.com/ethereum/go-ethereum/crypto/crypto.go:80 +0x98
  github.com/ethereum/go-ethereum/triedb/pathdb.(*reader).Node()
      /home/user/go/src/github.com/ethereum/go-ethereum/triedb/pathdb/reader.go:63 +0x11e
  github.com/ethereum/go-ethereum/trie.(*trieReader).node()
      /home/user/go/src/github.com/ethereum/go-ethereum/trie/trie_reader.go:69 +0x1a9
  github.com/ethereum/go-ethereum/trie.(*Trie).Prove()
```

Reply:
Fixed